rir/src/utils/FunctionHandle.h

```diff
@@ -48,23 +51,27 @@ class FunctionHandle {
         function->invocationCount = 0;
         function->markOpt = false;
 
-        return FunctionHandle(store);
+        FunctionHandle res(store);
+        res.storeOwner_ = true;
+
+        return res;
     }
 
-    FunctionHandle(SEXP store)
-        : store(store), payload(INTEGER(store)), capacity(XLENGTH(store)),
-          function((Function*)payload) {
+    explicit FunctionHandle(SEXP store)
+        : storeOwner_(false), store(store), payload(INTEGER(store)),
+          capacity(XLENGTH(store)), function((Function*)payload) {
         assert(function->magic == FUNCTION_MAGIC);
         assert(function->size <= capacity);
     }
 
     ~FunctionHandle() {
-        R_ReleaseObject(store);
+        if (storeOwner_)
+            R_ReleaseObject(store);
```

Comment:
so did you find out, why sometimes the handle is the owner and sometimes not? is that really a good design, or could it be improved at the client side?

for example, I could see an api, where you can either:
pass a store to the constructor   `FunctionHandle(store)` in which case it is *not* owned by the handle. Or let the handle create the store, eg. `FunctionHandle::create()`, in which case the store will be internal to the function handle and also owned by it.

Reply:
That is exactly what the code does, only the handle had no way to know in its destructor which way the store was obtained.

The problem I think was that the public constructor was used both implicitly (by the verifier and when retrieving the function from a `CodeHandle`) and explicitly (in the static `FunctionHandle::create` method). Hence I added the `explicit` modifier and made not owning the store the default. If the store is allocated by `create`, it changes the flag in the created handle and it later cleans up appropriately.

The design is ok imo if used in the way you described. You can create a handle from a SEXP and not care about the store, or if you need a new store, you use `create` and then have to take care to protect it after the handle is destroyed.